File: src/usb-manager/py/usb-manager.py
from subprocess import PIPE, Popen
from sys import stdout, platform, version_info
from os import system
from re import search
from time import sleep

import inquirer

from lib.utils import clear, spinning, cover, commandverify, usbverify, live_tasker, live_tasker_poweroff
from lib.printer import printer
from lib.reader import reader

LANGUAGE: int = 0

def core() -> None: clear(); language(); cover(); verify(); menu()

def language() -> None:

    global LANGUAGE

    print('Welcome / Bienvenido')

    q = [inquirer.List('language',
        message='Please, choose your language / Por favor selecciona tu idioma',
        choices=['English', 'Espanol']
    )]

    data = inquirer.prompt(q)
    if data['language'] == 'English': LANGUAGE = 1
    else: LANGUAGE = 2

def verify() -> None:

    if version_info < (3, 5):
        clear(); printer("error", 11, LANGUAGE); exit(1)
    if platform != "linux":
        clear(); printer("error",0, LANGUAGE); exit(1)
    if not commandverify("whiptail"):
        clear(); printer("error",3, LANGUAGE); exit(1)
    if not commandverify("udisksctl"):
        clear(); printer("error",2, LANGUAGE); exit(1)
    SERVICE: str = Popen('systemctl is-active udisks2', shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip('\n')
    if SERVICE == "inactive":
        clear(); printer("error",4, LANGUAGE); exit(1)
    usbverify(); printer("print",5, LANGUAGE)
    spinner = spinning()
    for _ in range(15):
        stdout.write(next(spinner))
        stdout.flush(); sleep(0.1)
        stdout.write('\b')

def menu() -> None:

    q = [inquirer.List('menu',
        message=reader(0, LANGUAGE),
        choices=[
            reader(1, LANGUAGE), reader(2, LANGUAGE),
            reader(3, LANGUAGE), reader(4, LANGUAGE)
        ]
    )]

    answers = inquirer.prompt(q)

    if answers['menu'] == reader(1, LANGUAGE): clear(); usblistener("mount")
    elif answers['menu'] == reader(2, LANGUAGE): clear(); usblistener("unmount")
    elif answers['menu'] == reader(3, LANGUAGE): clear(); usblistener("off")
    elif answers['menu'] == reader(4, LANGUAGE): clear(); exit(0)
    else: clear(); exit(0)

def usblistener(selector: str) -> None:

    clear(); usbverify()
    COUNT: int = 0; PARTS: list = []; BLOCK: list = []
    FLAGS: list = []; MOUNTS: list = []; USB: list = []
    UNMOUNTS: list = []; ARGS: list = []; ARGSPOWEROFF: list = []
    MOUNTCOUNT: int = 0; UNMOUNTCOUNT: int = 0; TYPE: str = ""
    MODEL: str = ""; FLAGSCOUNT: int = 0; DIRTYDEVS: list = []

    USB = Popen(r"""find /dev/disk/by-id/ -name 'usb*' | sort -n | sed 's/^\/dev\/disk\/by-id\///'
                    """, shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip().split('\n')
    for DEVICE in USB:
        DIRTYDEVS.append(Popen(f'readlink "/dev/disk/by-id/{DEVICE}"', shell=True,
                            stdout=PIPE).stdout.read().decode('utf-8').rstrip().split("\n")[0])

    DIRTYDEVS = list(filter(('').__ne__, DIRTYDEVS))
    for DEV in DIRTYDEVS:
        ABSOLUTEPARTS = Popen(f"""
                            echo {DEV} | sed 's/^\.\.\/\.\.\//\/dev\//' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'""",
                            shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()

        if ABSOLUTEPARTS != "":
            PARTS.append(Popen(f"""
                                echo {DEV} | sed 's/^\.\.\/\.\.\///' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'""",
                                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip().split("\n")[0])
            COUNT += 1
        else:
            BLOCK.append(Popen(f"echo {DEV} | sed 's/^\.\.\/\.\.\///'", shell=True, stdout=PIPE).stdout.read()
                            .decode('utf-8').rstrip().split("\n")[0])
            FLAGS.append(COUNT)

    for PARTITIONS in PARTS:
        MOUNTED: str = Popen(f"lsblk /dev/{PARTITIONS} | sed -ne '/\//p'", shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
        if MOUNTED != "": UNMOUNTCOUNT += 1; MOUNTS.append(PARTITIONS)
        else: MOUNTCOUNT += 1; UNMOUNTS.append(PARTITIONS)

    if selector == "mount":
        if UNMOUNTCOUNT == COUNT: printer("error",7); input(reader(5)); menu(); return
        COUNT=0
        for PART in UNMOUNTS:
            DEVICE: str = f"/dev/{PART}"
            TYPE: str = Popen(r'''
                lsblk -f /dev/'''+PART+''' | sed -ne '2p' | cut -d " " -f2''',
                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
            try: TEMP = FLAGS[FLAGSCOUNT]
            except IndexError: pass
            if COUNT == TEMP :
                try: BLOCKSTAT = BLOCK[FLAGSCOUNT]
                except IndexError: pass
                FLAGSCOUNT += 1
            MODEL: str = Popen(f'''
                cat /sys/class/block/{BLOCKSTAT}/device/model''',
                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
            ARGS.append([DEVICE, MODEL + " " + TYPE]); COUNT+=1

        q = [inquirer.List('menu',
            message=reader(6, LANGUAGE),
            choices=ARGS
        )]

        answers = inquirer.prompt(q)
        clear(); mountaction(answers['menu'])

    elif selector == "unmount":
        if MOUNTCOUNT == COUNT: printer("error",10); input(reader(5)); menu(); return
        COUNT=0
        for PART in MOUNTS:
            DEVICE: str = "/dev/"+PART
            TYPE: str = Popen(r'''
                lsblk -f /dev/'''+PART+''' | sed -ne '2p' | cut -d " " -f2''',
                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
            try: TEMP = FLAGS[FLAGSCOUNT]
            except IndexError: pass
            if COUNT == TEMP:
                try: BLOCKSTAT = BLOCK[FLAGSCOUNT]
                except IndexError: pass
                FLAGSCOUNT += 1
            MODEL: str = Popen(f'''
                cat /sys/class/block/{BLOCKSTAT}/device/model''',
                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
            ARGS.append([DEVICE, MODEL + " " + TYPE]); COUNT +=1

        q = [inquirer.List('menu',
            message=reader(9, LANGUAGE),
            choices=ARGS
        )]

        answers = inquirer.prompt(q)
        clear(); unmountaction(answers['menu'])

    elif(selector == "off"):
        COUNT=0
        for PART in BLOCK:
            DEVICE: str = "/dev/"+PART
            BLOCKSTAT = BLOCK[COUNT]
            MODEL: str = Popen(f'''
                cat /sys/class/block/{BLOCKSTAT}/device/model''',
                shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()
            ARGSPOWEROFF.append([DEVICE, MODEL]); COUNT += 1
        q = [inquirer.List('menu',
            message=reader(11, LANGUAGE),
            choices=ARGS
        )]

        answers = inquirer.prompt(q)
        clear(); poweroffaction(answers['menu'])

def mountaction(part: str) -> None:

    clear()
    if part == "": return

    capture = live_tasker(f"udisksctl mount -b {part}")

    if capture[0] == 0:
        system(f"whiptail --title '{reader(8)}' --msgbox '{reader(8)+capture[1]}' 7 35")
        menu()
    else:
        if search("NotAuthorized*",capture[1]):
            clear(); printer("error",8); input(reader(5)); menu()
        elif search("NotAuthorizedDismissed*",capture[1]):
            clear(); printer("error",8); input(reader(5)); menu()
        elif search("AlreadyMounted*",capture[1]):
            clear(); printer("error",9); input(reader(5)); menu()

def unmountaction(part: str) -> None:

    clear()
    if part == "": return

    capture = live_tasker(f"udisksctl unmount -b {part}")

    if capture[0] == 0:
        system(f"whiptail --title '{reader(8)}' --msgbox '{reader(8)+capture[1]}' 7 35")
        menu()
    else:
        if search("NotAuthorized*",capture[1]):
            clear(); printer("error",8); input(reader(5)); menu()
        elif search("NotAuthorizedDismissed*",capture[1]):
            clear(); printer("error",8); input(reader(5)); menu()
        elif search("AlreadyMounted*",capture[1]):
            clear(); printer("error",9); input(reader(5)); menu()

def poweroffaction(part: str) -> None:

    clear()
    if part == "": return

    BLOCKTEMP: str = Popen(f'echo {part} | cut -d "/" -f3',
                            shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()

    PARTITIONSQUERY: list = Popen("""find /dev -name \"""" +BLOCKTEMP+
                                    """[[:digit:]]\" | sort -n | sed 's/^\/dev\///'""", shell=True,
                                stdout=PIPE).stdout.read().decode('utf-8').rstrip().split("\n")

    for PARTITION in PARTITIONSQUERY:
        capture: list = live_tasker(f"udisksctl unmount -b {PARTITION} &> /dev/null")
        if capture[0] == 0:
            spinner = spinning()
            for _ in range(10):
                stdout.write(next(spinner))
                stdout.flush(); sleep(0.1)
                stdout.write('\b')
        else:
            if LANGUAGE == 1:
                system(f"whiptail --title 'ERROR' --msgbox 'FAIL: Error unmounting /dev/{PARTITION} please check or check if you have the right permissions' 7 35")
                menu(); return
            else:
                system(f"whiptail --title 'ERROR' --msgbox 'ERROR: Hubo un error desmontando /dev/{PARTITION} por favor revisar o mira si tienes permisos' 7 35")
                menu(); return

    MODEL: str = Popen(f'cat /sys/class/block/{BLOCKTEMP}/device/model',
                        shell=True, stdout=PIPE).stdout.read().decode('utf-8').rstrip()

    capture: int = live_tasker_poweroff(f"udisksctl power-off -b {part}")

    if capture == 0:
        if LANGUAGE == 1:
            system(f"whiptail --title 'SUCCESS' --msgbox 'SUCCESS: Your device {MODEL} was succesfully power-off' 7 35")
            menu()
        else:
            system(f"whiptail --title 'LISTO' --msgbox 'LISTO: Tu dispositivo {MODEL} se ha apagado exitosamente' 7 35")
            menu()
    else:
        if LANGUAGE == 1:
            system(f"whiptail --title 'FAIL' --msgbox 'FAIL: Power-off is not available on this device, please check or check if you have permissions' 7 35")
            menu(); return
        else:
            system(f"whiptail --title 'ERROR' --msgbox 'ERROR: no está disponible el apagar este dispositivo, por favor revisar o mira si tienes permisos' 7 35")
            menu(); return

if __name__ == "__main__": core()

File: src/go/usb-manager/usb-manager.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strings"
	"time"
	"usb-manager/lib"

	"github.com/AlecAivazis/survey/v2"
	"github.com/briandowns/spinner"
)

func core() {
	lib.Clear(); language(); lib.Cover(); verify()
}

var LANGUAGE int = 0

func language() {

	option := 0
	promptLang := &survey.Select{
		Message: "Bienvenido / Welcome \n Choose your language / Selecciona tu idioma",
		Options: []string{"English", "Espanol"},
	}
	survey.AskOne(promptLang, &option)

	if option == 0 {
		LANGUAGE = 1
	} else {
		LANGUAGE = 2
	}
}

func usbverify() {
	SVERIFY,_ := exec.Command("bash", "-c", "find /dev/disk/by-id/ -name 'usb*' | sort -n | sed 's/^\\/dev\\/disk\\/by-id\\///'").Output()
	if string(SVERIFY) == "" {
		lib.Clear()
		lib.Printer("error", 6, LANGUAGE)
		fmt.Println(""); os.Exit(1)
	}
}

func verify() {
	platform := runtime.GOOS
	if platform != "linux" {
		lib.Clear(); lib.Printer("error", 0, LANGUAGE)
		fmt.Print("\n"); os.Exit(1)
	}
	if !lib.Commandverify("whiptail") {
		lib.Clear(); lib.Printer("error", 3, LANGUAGE)
		fmt.Print("\n"); os.Exit(1)
	}
	if !lib.Commandverify("udisksctl") {
		lib.Clear(); lib.Printer("error", 2, LANGUAGE)
		fmt.Print("\n"); os.Exit(1)
	}

	SSYSUD, _  := exec.Command("bash", "-c","systemctl is-active udisks2").Output()
	if(string(SSYSUD) == "inactive") {
		lib.Clear(); lib.Printer("error", 4, LANGUAGE)
		fmt.Println(); os.Exit(1)
	}

	usbverify()
	lib.Printer("print", 5, LANGUAGE)
	s := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
	s.Start(); time.Sleep(time.Second); s.Stop()

	lib.Clear(); menu()
}

func menu() {
	option := 0
	promptMenu := &survey.Select{
		Message: lib.Reader(0, LANGUAGE),
		Options: []string{
			lib.Reader(1, LANGUAGE), lib.Reader(2, LANGUAGE),
			lib.Reader(3, LANGUAGE), lib.Reader(4, LANGUAGE),
		},
	}
	survey.AskOne(promptMenu, &option)

	switch option {
		case 0:
			lib.Clear(); usblistener("mount")
		case 1:
			lib.Clear(); usblistener("unmount")
		case 2:
			lib.Clear(); usblistener("off")
		case 3:
			lib.Clear(); os.Exit(0)
		default:
			lib.Clear(); os.Exit(0)
	}
}

func usblistener(selector string) {

	COUNT := 0
	PARTS, USB := make([]string, 20), make([]string, 20)
	BLOCK, FLAGS := make([]string, 20), make([]int, 0,20)
	MOUNTS, UNMOUNTS := make([]string, 20), make([]string, 20)
	DIRTYDEVS, ARGSPOWEROFF := make([]string, 20), make([]string, 20)
	MOUNTCOUNT, UNMOUNTCOUNT := 0,0;
	FLAGSCOUNT, ARGS := 0, make([]string, 20)
	lib.Clear(); usbverify()

	SUSB, _ := exec.Command("bash", "-c", "find /dev/disk/by-id/ -name 'usb*' | sort -n | sed 's/^\\/dev\\/disk\\/by-id\\///'").Output()
	USB = strings.Split(string(SUSB), "\n")

	for _, DEVICE := range USB {
		SDIRTDEV, _ := exec.Command("bash", "-c", fmt.Sprintf("readlink \"/dev/disk/by-id/%s\"",DEVICE)).Output()
		DIRTYDEVS = append(DIRTYDEVS, strings.Split(string(SDIRTDEV), "\n")[0])
	}

	DIRTYDEVS = lib.RemoveWhere(DIRTYDEVS,"")

	for _, DEV  := range DIRTYDEVS {
		SABSPARTS, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | sed 's/^\\.\\.\\/\\.\\.\\//\\/dev\\//' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'", DEV)).Output()
		ABSOLUTEPARTS := strings.Replace(string(SABSPARTS), "\n", "", 1)
		if ABSOLUTEPARTS != "" {
			SPARTS, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | sed 's/^\\.\\.\\/\\.\\.\\///' | sed '/.*[[:alpha:]]$/d' | sed '/blk[[:digit:]]$/d'", DEV)).Output()
			PARTS = append(PARTS, strings.Split(string(SPARTS), "\n")[0])
			COUNT++
		} else {
			SBLOCK, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | sed 's/^\\.\\.\\/\\.\\.\\///'", DEV)).Output()
			BLOCK = append(BLOCK, strings.Split(string(SBLOCK), "\n")[0])
			FLAGS = append(FLAGS, COUNT)
		}
	}

	PARTS = lib.RemoveWhere(PARTS,"")
	BLOCK = lib.RemoveWhere(BLOCK,"")

	for _, PARTITIONS  := range PARTS {
		SMOUNTED, _ := exec.Command("bash", "-c", fmt.Sprintf("lsblk /dev/%s | sed -ne '/\\//p'", PARTITIONS)).Output()
		MOUNTED := strings.Replace(string(SMOUNTED), "\n", "", 1)
		if MOUNTED != "" {
			UNMOUNTCOUNT++
			MOUNTS = append(MOUNTS, PARTITIONS)
			} else {
			MOUNTCOUNT++
			UNMOUNTS = append(UNMOUNTS, PARTITIONS)
		}
	}

	MOUNTS = lib.RemoveWhere(MOUNTS,"")
	UNMOUNTS = lib.RemoveWhere(UNMOUNTS,"")

	if(selector == "mount") {
		if(UNMOUNTCOUNT == COUNT) {
			lib.Clear(); lib.Printer("error", 7, LANGUAGE)
			fmt.Println("")
			fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
			return
		}

		COUNT = 0

		for _, PART  := range UNMOUNTS {
			DEVICE := fmt.Sprintf("/dev/%s", PART)
			TEMP := FLAGS[FLAGSCOUNT]
			if(COUNT == TEMP){
				FLAGSCOUNT++
			}
			ARGS = append(ARGS, DEVICE)
			COUNT++
		}

		ARGS = lib.RemoveWhere(ARGS,"")
		ARGS = append(ARGS, lib.Reader(16, LANGUAGE))

		option := 0
		selectionMenu := &survey.Select{
			Message: lib.Reader(6, LANGUAGE),
			Options: ARGS,
		}
		survey.AskOne(selectionMenu, &option)

		if(ARGS[option] == lib.Reader(16, LANGUAGE)) {
			lib.Clear(); menu()
		} else {
			lib.Clear(); mountAction(ARGS[option])
		}

	} else if(selector == "unmount") {
		if(MOUNTCOUNT == COUNT) {
			lib.Clear(); lib.Printer("error", 10, LANGUAGE)
			fmt.Println("")
			fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
			return
		}

		COUNT = 0

		for _, PART  := range MOUNTS {
			DEVICE := fmt.Sprintf("/dev/%s", PART)
			TEMP := FLAGS[FLAGSCOUNT]
			if(COUNT == TEMP){
				FLAGSCOUNT++
			}
			ARGS = append(ARGS, DEVICE)
			COUNT++
		}

		ARGS = lib.RemoveWhere(ARGS,"")
		ARGS = append(ARGS, lib.Reader(16, LANGUAGE))

		option := 0
		selectionMenu := &survey.Select{
			Message: lib.Reader(9, LANGUAGE),
			Options: ARGS,
		}
		survey.AskOne(selectionMenu, &option)

		if(ARGS[option] == lib.Reader(16, LANGUAGE)) {
			lib.Clear(); menu()
		} else {
			lib.Clear(); unMountAction(ARGS[option])
		}
	} else if(selector == "off") {
		if(MOUNTCOUNT == COUNT) {
			lib.Clear(); lib.Printer("error", 10, LANGUAGE)
			fmt.Println("")
			fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
			return
		}

		COUNT = 0

		for _, PART := range BLOCK {
			DEVICE := fmt.Sprintf("/dev/%s", PART)
			ARGSPOWEROFF = append(ARGSPOWEROFF, DEVICE)
			COUNT++
		}

		ARGSPOWEROFF = lib.RemoveWhere(ARGSPOWEROFF,"")
		ARGSPOWEROFF = append(ARGSPOWEROFF, lib.Reader(16, LANGUAGE))

		option := 0
		selectionMenu := &survey.Select{
			Message: lib.Reader(11, LANGUAGE),
			Options: ARGSPOWEROFF,
		}
		survey.AskOne(selectionMenu, &option)

		if(ARGS[option] == lib.Reader(16, LANGUAGE)) {
			lib.Clear(); menu()
		} else {
			lib.Clear(); powerOffAction(ARGSPOWEROFF[option])
		}
	}
}

func mountAction(part string) {
	lib.Clear()
	if(part == "") {
		return;
	}
	SMOUNTACTION, EMOUNTACTION := exec.Command("bash", "-c", fmt.Sprintf("udisksctl mount -b %s", part)).Output()
	if(EMOUNTACTION == nil) {
		dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title '%s' --msgbox '%s %s' 7 60",lib.Reader(8, LANGUAGE),lib.Reader(8, LANGUAGE), string(SMOUNTACTION)))
		dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
		_ = dialog.Run()
		lib.Clear(); menu()
	} else {
		reg1Bool, _ := regexp.MatchString("NotAuthorized*", string(SMOUNTACTION))
		reg2Bool, _ := regexp.MatchString("NotAuthorizedDismissed*", string(SMOUNTACTION))
		reg3Bool, _ := regexp.MatchString("AlreadyMounted*", string(SMOUNTACTION))
		if(reg1Bool) {
			lib.Clear(); lib.Printer("error", 8, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		} else if(reg2Bool) {
			lib.Clear(); lib.Printer("error", 8, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		} else if(reg3Bool) {
			lib.Clear(); lib.Printer("error", 9, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		}
	}
}

func unMountAction(part string) {
	lib.Clear()
	if(part == "") {
		return;
	}
	SMOUNTACTION, EMOUNTACTION := exec.Command("bash", "-c", fmt.Sprintf("udisksctl unmount -b %s", part)).Output()
	if(EMOUNTACTION == nil) {
		dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title '%s' --msgbox '%s %s' 7 60",lib.Reader(8, LANGUAGE),lib.Reader(8, LANGUAGE), string(SMOUNTACTION)))
		dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
		_ = dialog.Run()
		lib.Clear(); menu()
	} else {
		reg1Bool, _ := regexp.MatchString("NotAuthorized*", string(SMOUNTACTION))
		reg2Bool, _ := regexp.MatchString("NotAuthorizedDismissed*", string(SMOUNTACTION))
		reg3Bool, _ := regexp.MatchString("AlreadyMounted*", string(SMOUNTACTION))
		if(reg1Bool) {
			lib.Clear(); lib.Printer("error", 8, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		} else if(reg2Bool) {
			lib.Clear(); lib.Printer("error", 8, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		} else if(reg3Bool) {
			lib.Clear(); lib.Printer("error", 9, LANGUAGE)
			fmt.Println(); fmt.Println(lib.Reader(5, LANGUAGE))
			fmt.Scanln(); lib.Clear(); menu()
		}
	}
}

func powerOffAction(part string) {
	lib.Clear()
	if(part == "") {
		return;
	}

	SBLOCKTEMP, _ := exec.Command("bash", "-c", fmt.Sprintf("echo %s | cut -d \"/\" -f3",part)).Output()
	BLOCKTEMP := strings.Replace(string(SBLOCKTEMP), "\n", "", 1)

	SPARTITIONS, _ := exec.Command("bash", "-c", fmt.Sprintf("find /dev -name \"%s[[:digit:]]\" | sort -n | sed 's/^\\/dev\\///' ",BLOCKTEMP)).Output()
	PARTITIONS := strings.Split(string(SPARTITIONS), "\n")

	PARTITIONS = lib.RemoveWhere(PARTITIONS, "")

	for _, PARTITION  := range PARTITIONS {
		_, EUNMOUNT := exec.Command("bash", "-c", fmt.Sprintf("udisksctl unmount -b /dev/%s &> /dev/null ",PARTITION)).Output()
		if(EUNMOUNT == nil){
			s := spinner.New(spinner.CharSets[9], 100*time.Millisecond)
			s.Start(); time.Sleep(time.Second); s.Stop()
		} else {
			if(LANGUAGE == 1) {
				dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'ERROR' --msgbox 'FAIL: Error unmounting /dev/%s please check or check if you have the right permissions' 7 60",PARTITION))
				dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
				_ = dialog.Run()
				lib.Clear(); menu()
				return
			} else {
				dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'ERROR' --msgbox 'ERROR: Hubo un error desmontando /dev/%s por favor revisar o mira si tienes permisos 7 60",PARTITION))
				dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
				_ = dialog.Run()
				lib.Clear(); menu()
				return
			}
		}
	}

	SMODEL, _ := exec.Command("bash", "-c", "cat /sys/class/block/%s/device/model",BLOCKTEMP).Output()
	MODEL := strings.Replace(string(SMODEL), "\n", "", 1)

	_, EPOWEROFF := exec.Command("bash", "-c", fmt.Sprintf("udisksctl power-off -b %s ",part)).Output()
	if(EPOWEROFF == nil) {
		if(LANGUAGE == 1) {
			dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'SUCCESS' --msgbox 'SUCCESS: Your device %s was succesfully power-off' 7 60",MODEL))
			dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
			_ = dialog.Run()
			lib.Clear(); menu()
			return
		} else {
			dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'LISTO' --msgbox 'LISTO: Tu dispositivo %s se ha apagado exitosamente' 7 60",MODEL))
			dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
			_ = dialog.Run()
			lib.Clear(); menu()
			return
		}
	} else {
		if(LANGUAGE == 1) {
			dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'FAIL' --msgbox 'FAIL:  Power-off is not available on this device, please check or check if you have permissions' 7 60"))
			dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
			_ = dialog.Run()
			lib.Clear(); menu()
			return
		} else {
			dialog := exec.Command("bash", "-c",fmt.Sprintf("whiptail --title 'ERROR: no está disponible el apagar este dispositivo, por favor revisar o mira si tienes permisos' 7 60"))
			dialog.Stdin = os.Stdin; dialog.Stdout = os.Stdout; dialog.Stderr = os.Stderr
			_ = dialog.Run()
			lib.Clear(); menu()
			return
		}
	}


}

func main() { core() }